src/zarr/v3/group.py

```diff
@@ -150,15 +158,25 @@ def from_dict(
     async def getitem(
         self,
         key: str,
-    ) -> Union[AsyncArray, AsyncGroup]:
+    ) -> AsyncArray | AsyncGroup:
 
         store_path = self.store_path / key
 
+        # Note:
+        # in zarr-python v2, we first check if `key` references an Array, else if `key` references
+        # a group,using standalone `contains_array` and `contains_group` functions. These functions
+        # are reusable, but for v3 they would perform redundant I/O operations.
+        # Not clear how much of that strategy we want to keep here.
+
+        # if `key` names an object in storage, it cannot be an array or group
+        if await store_path.exists():
+            raise KeyError(key)
```

Comment:
I don't think we should include this check. This adds a roundtrip to the store to protect against a situation I have a hard time imaging being a real issue.


For one, most storage backends forbid such a situation but I don't think the spec really cares.

```
/store/a/b/c
/store/a/b/c/zarr.json
```

Reply:
I opened https://github.com/zarr-developers/zarr-specs/pull/291 to discuss removing implicit groups from the spec. the PR is at https://github.com/zarr-developers/zarr-specs/pull/292